### PR TITLE
Documentation for "Immutable Secrets/ConfigMaps" feature

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/enhancements/issues/1412

/assign @cheftako

### CODE REVIEW

**Comment by wojtek-t** [comment]
> /sig apps

This isn't sig apps - it's sig/storage with combination of sig/node

File: content/en/docs/concepts/configuration/secret.md [hunk 1]
```diff
@@ -672,6 +672,16 @@ A container using a Secret as a
 Secret updates.
 {{< /note >}}
 
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+For clusters that are extensively using secrets (having at least tens of thousands of unique
+secret to pod mounts), we designed the `Immutable Secrets/ConfigMaps` feature. It allows you
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: done

File: content/en/docs/concepts/configuration/secret.md [hunk 2]
```diff
@@ -672,6 +672,16 @@ A container using a Secret as a
...
+to mark individual Secrets/ConfigMaps as `Immutable`, which blocks any changes to their data
+and in return:
+- protects you from accidental bad updates that could cause applications outages
+- improves performance of your cluster by significantly reducing load on kube-apiserver, by
+  closing watches for secrets marked as immutable
+
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: Added (btw - it's not possible to change the data once mark as immutable - even by priviliged user).

[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md]: > it's not possible to change the data once mark as immutable - even by priviliged user

If I really want to make that change, it sounds like I'd need to delete the cluster and recreate it, then? Or can I delete the Secret and recreate it?

[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md]: :thought_balloon: are there any security implications, for example if a user is able to mark a ConfigMap immutable and thereby break a process that expects to be able to modify it?

[the remark this answers is not in the document]
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: > Or can I delete the Secret and recreate it?

You can delete the secret and create a new one, but that won't be picked by pods that are mounting it - you will also need to recreate those pods.

> are there any security implications, for example if a user is able to mark a ConfigMap immutable and thereby break a process that expects to be able to modify it?

If a user is able to modify a config map that automation should only modify, it can already break stuff, but setting malicious data in it.

[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md]: > You can delete the secret and create a new one, but that won't be picked by pods that are mounting it - you will also need to recreate those pods.

That's the sort of detail I think readers would be interested in.

[the remark this answers is not in the document]
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: Added - thanks!

File: content/en/docs/concepts/configuration/secret.md [hunk 3]
```diff
@@ -672,6 +672,33 @@ A container using a Secret as a
...
 {{< /note >}}
 
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+For clusters that are extensively using secrets (having at least tens of thousands of unique
+Secret to Pod mounts), Kubernetes has an _Immutable Secrets/ConfigMaps_ feature. It allows you
+to mark individual Secrets/ConfigMaps as `Immutable`, which blocks any changes to their data
+and in return:
+- protects you from accidental bad updates that could cause applications outages
+- improves performance of your cluster by significantly reducing load on kube-apiserver, by
```
Comment (kbhawkey) [inline content/en/docs/concepts/configuration/secret.md]: Hello @wojtek-t ,
I wonder if a new commit would fix the deploy preview. Could you add a new commit?
line 681, Possible edit: `- protects you from accidental (or unwanted) updates
line 699, spelling, secret

/cc @VineethReddy02

File: content/en/docs/concepts/configuration/secret.md [hunk 4]
```diff
@@ -672,6 +672,33 @@ A container using a Secret as a
...
+  closing watches for secrets marked as immutable
+
+To use it, you need to enable `ImmutableEmphemeralVolumes`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) and then
+mark your Secret/ConfigMap by setting it `Immutable` field to true, e.g.:
```
Comment (sftim) [inline content/en/docs/concepts/configuration/secret.md]: nits:
```suggestion
mark your Secret/ConfigMap by setting its `Immutable` field to true; for example:
```

File: content/en/docs/concepts/configuration/secret.md [hunk 5]
```diff
@@ -672,6 +672,33 @@ A container using a Secret as a
...
+kind: Secret
+metadata:
+  ...
+data:
+  ...
+immutable: true
+```
+
+Once marked as immutable, it is NOT possible to revert this change nor to mutate the contents
+of data field. You can only delete and recreate the secrete, but that will not be picked by
```
Comment (sftim) [inline content/en/docs/concepts/configuration/secret.md]: nit:
```suggestion
of data field. You can only delete and recreate the Secret, but that will not be picked by
```

Also, I'd put this in a `{{< note >}}` or `{{< caution >}}` block to highlight the gotcha about Pod replacement.

File: content/en/docs/concepts/configuration/secret.md (hunk 3 above)
Comment (kbhawkey) [reply content/en/docs/concepts/configuration/secret.md]: @wojtek-t , Would you rebase this PR? I am interested in viewing a preview of the page.
You may have picked up a previous version of the `zh` deployment.md file. The build error has been corrected.

File: content/en/docs/concepts/configuration/secret.md (hunk 5 above)
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: done

File: content/en/docs/concepts/configuration/secret.md (hunk 4 above)
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: done

File: content/en/docs/concepts/configuration/secret.md (hunk 3 above)
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: done

**Comment by VineethReddy02** [comment]
👋 Please rebase this PR on `dev-1.18` in order to avoid merge conflicts.

Feel free to `/hold cancel` after you've rebased.

/hold

File: content/en/docs/concepts/configuration/secret.md [hunk 6]
```diff
@@ -672,6 +672,35 @@ A container using a Secret as a
 Secret updates.
 {{< /note >}}
 
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+For clusters that are extensively using secrets (having at least tens of thousands of unique
+Secret to Pod mounts), Kubernetes has an _Immutable Secrets/ConfigMaps_ feature. It allows you
+to mark individual Secrets/ConfigMaps as `Immutable`, which blocks any changes to their data
+and in return:
```
Comment (sftim) [inline content/en/docs/concepts/configuration/secret.md]: Can I use an immutable Secret if I only have a few hundred Secrets in my cluster?

File: content/en/docs/concepts/configuration/secret.md [hunk 7]
```diff
@@ -672,6 +672,35 @@ A container using a Secret as a
...
+- protects you from accidental (or unwanted) updates that could cause applications outages
+- improves performance of your cluster by significantly reducing load on kube-apiserver, by
+  closing watches for secrets marked as immutable
+
+To use it, you need to enable `ImmutableEmphemeralVolumes`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) and then
+mark your Secret/ConfigMap by setting its `Immutable` field to true, e.g.:
```
Comment (sftim) [inline content/en/docs/concepts/configuration/secret.md]: nit:
```suggestion
mark your Secret or ConfigMap, by setting its `immutable` field to `true`;
for example:
```

File: content/en/docs/concepts/configuration/secret.md (hunk 6 above)
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: Yes - but I don't see where the text suggests you can't.

File: content/en/docs/concepts/configuration/secret.md (hunk 7 above)
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: done

File: content/en/docs/concepts/configuration/secret.md (hunk 6 above)
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md]: > For clusters that are extensively using secrets (having at least tens of thousands of unique Secret to Pod mounts), Kubernetes has an _Immutable Secrets/ConfigMaps_ feature.

This could read like a requirement.

File: content/en/docs/concepts/configuration/secret.md [hunk 8]
```diff
@@ -676,6 +676,35 @@ A container using a Secret as a
...
 
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+For clusters that are extensively using secrets (having at least tens of thousands of unique
+Secret to Pod mounts), Kubernetes has an _Immutable Secrets/ConfigMaps_ feature. It allows you
+to mark individual Secrets/ConfigMaps as `Immutable`, which blocks any changes to their data
+and in return:
+- protects you from accidental (or unwanted) updates that could cause applications outages
+- improves performance of your cluster by significantly reducing load on kube-apiserver, by
+  closing watches for secrets marked as immutable
```
Comment (sftim) [inline content/en/docs/concepts/configuration/secret.md]:
```suggestion
closing watches for secrets marked as immutable.
```

Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: done

File: content/en/docs/concepts/configuration/secret.md (hunk 6 above)
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: What do you suggest? [I don't read it as a requirement]

**Comment by wojtek-t** [comment]
@kbhawkey - can you please take another pass (since we already have a technical lgtm now).

**Comment by kbhawkey** [comment]
👀

**Comment by kbhawkey** [comment]
Page previews:
https://deploy-preview-19297--kubernetes-io-vnext-staging.netlify.com/docs/concepts/configuration/secret/

https://deploy-preview-19297--kubernetes-io-vnext-staging.netlify.com/docs/reference/command-line-tools-reference/feature-gates/

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 9]
```diff
@@ -105,6 +105,7 @@ different Kubernetes components.
 | `HPAScaleToZero` | `false` | Alpha | 1.16 | |
 | `HugePageStorageMediumSize` | `false` | Alpha | 1.18 | |
 | `HyperVContainer` | `false` | Alpha | 1.10 | |
+| `ImmutableEphemeralVolumes` | `false` | Alpha | 1.18 | |
```
Comment (kbhawkey) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md]: @wojtek-t , If time, you probably want to add a brief description about this feature.
See the `feature-gates` section on this page, https://deploy-preview-19297--kubernetes-io-vnext-staging.netlify.com/docs/reference/command-line-tools-reference/feature-gates/#feature-gates

File: content/en/docs/concepts/configuration/secret.md [hunk 10]
```diff
@@ -676,6 +676,35 @@ A container using a Secret as a
 Secret updates.
 {{< /note >}}
 
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
```
Comment (kbhawkey) [inline content/en/docs/concepts/configuration/secret.md]: @wojtek-t , Here are some suggestions.
Could you create a sub-heading to introduce this section?
Right now this content is under `Using Secrets` and `mounted-secrets-are-updated-automatically`.
Not sure if this makes sense?
````
#### Immutable Secrets and ConfigMaps
````

Possible edit to introduce feature:

The Kubernetes alpha feature _Immutable Secrets and ConfigMaps_ provides an option to set individual Secrets and ConfigMaps as immutable. For clusters that extensively use secrets (at least tens of thousands of unique Secret to Pod mounts), preventing changes to this data has the following advantages:

File: content/en/docs/concepts/configuration/secret.md [hunk 11]
```diff
@@ -676,6 +676,35 @@ A container using a Secret as a
...
+
+For clusters that are extensively using secrets (having at least tens of thousands of unique
+Secret to Pod mounts), Kubernetes has an _Immutable Secrets/ConfigMaps_ feature. It allows you
+to mark individual Secrets/ConfigMaps as `Immutable`, which blocks any changes to their data
+and in return:
+- protects you from accidental (or unwanted) updates that could cause applications outages
+- improves performance of your cluster by significantly reducing load on kube-apiserver, by
+closing watches for secrets marked as immutable.
```
Comment (kbhawkey) [inline content/en/docs/concepts/configuration/secret.md]: nit, Line 686, `closing watches`?
- improves performance of your cluster by significantly reducing load on the kube-apiserver and closing watches for secrets marked as immutable.

File: content/en/docs/concepts/configuration/secret.md [hunk 12]
```diff
@@ -676,6 +676,35 @@ A container using a Secret as a
...
+
+To use it, you need to enable `ImmutableEmphemeralVolumes`
```
Comment (kbhawkey) [inline content/en/docs/concepts/configuration/secret.md]: Possible edit, something like:

To use this feature, enable the [ImmutableEmphemeralVolumes feature gate](...)
and set your Secret or ConfigMap `immutable` field to `true`. For example:

File: content/en/docs/concepts/configuration/secret.md [hunk 13]
```diff
@@ -676,6 +676,35 @@ A container using a Secret as a
...
+metadata:
+  ...
+data:
+  ...
+immutable: true
+```
+
+{{< note >}}
+Once marked as immutable, it is NOT possible to revert this change nor to mutate the contents
+of data field. You can only delete and recreate the Secret, but that will not be picked by
```
Comment (kbhawkey) [inline content/en/docs/concepts/configuration/secret.md]: Possible edit, pick and choose what makes sense:
Once a Secret is marked as immutable, it is *not* possible to revert this change nor to mutate the contents of the `data` field. You can only delete and recreate the Secret. Existing Pods maintain a mount point to the deleted Secret. It is recommended to recreate these Pods.

File: content/en/docs/concepts/configuration/secret.md (hunk 12 above)
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: done

File: content/en/docs/concepts/configuration/secret.md (hunk 11 above)
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: I think that "by" is better - because the performance is improved by the fact that we're closing those watches.

File: content/en/docs/concepts/configuration/secret.md (hunk 13 above)
Comment (wojtek-t) [reply content/en/docs/concepts/configuration/secret.md]: done

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 9 above)
Comment (wojtek-t) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md]: done

File: content/en/docs/concepts/configuration/secret.md [hunk 14]
```diff
@@ -676,6 +676,36 @@ A container using a Secret as a
 Secret updates.
 {{< /note >}}
 
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+The Kubernetes alpha feature _Immutable Secrets and ConfigMaps_ provides an option to set
+individual Secrets and ConfigMaps as immutable. For clusters that extensively use Secrets
+(at least tens of thousands of unique Secret to Pod mounts), preventing changes to their
+data has the following advantages:
+- protects you from accidental (or unwanted) updates that could cause applications outages
```
Comment (kbhawkey) [inline content/en/docs/concepts/configuration/secret.md]: @wojtek-t , Changes look good. One issue:
Add a newline at line 685 before the first list item (the list is not rendering)
Thanks!

**Comment by wojtek-t** [comment]
@kbhawkey - done (hopefully will work this time)

**Comment by kbhawkey** [comment]
/lgtm

@VineethReddy02 , Is the `dev-1.18` branch accepting commits?

**Comment by VineethReddy02** [comment]
@kbhawkey yes! If the technical reviews are done, we can merge this into dev-1.18

**Comment by wojtek-t** [comment]
@saad-ali  approved it yesterday (there were only wording changes requested by @kbhawkey after that, so I believe we're fine).

**Comment by wojtek-t** [comment]
@kbhawkey - can you approve it then?

**Comment by kbhawkey** [comment]
@wojtek-t , I assume from the changes that you have rebased this pull request a number of times (there was a mixup with the dev-1.18 branch early in the release). If so, @VineethReddy02, I am happy to approve this pull request or if you agree, you can as well.

**Comment by wojtek-t** [comment]
> @wojtek-t , I assume from the changes that you have rebased this pull request a number of times (there was a mixup with the dev-1.18 branch early in the release). If so, @VineethReddy02, I am happy to approve this pull request or if you agree, you can as well.

Yeah - I think it's ready. I would feel better if you could approve it.

**Comment by sftim** [comment]
Based on https://github.com/kubernetes/website/pull/19297#issuecomment-598398147 and https://github.com/kubernetes/website/pull/19297#issuecomment-598713641
/approve

**Comment by wojtek-t** [comment]
Thanks!
